navigate-route/src/main/java/com/esri/arcgismaps/sample/navigateroute/MainActivity.kt

```diff
@@ -0,0 +1,405 @@
+                // switch to the next destination
+                routeTracker.switchToNextDestination().getOrElse {
+                    return@launch showError("Error retrieving next destination: ${it.message}")
+                }
+                // set second stop message
+                nextStopTextView.text = resources.getStringArray(R.array.stop_message)[1]
+            } else {
+                // the final destination has been reached,
+                // stop the simulated location data source
+                simulatedLocationDataSource.stop()
+                routeTrackerLocationDataSource.stop()
+                // set last stop message
+                nextStopTextView.text = resources.getStringArray(R.array.stop_message)[2]
+            }
+        }
+    }
+
+    /**
+     * Update the remaining and traveled route graphics using [trackingStatus]
+     */
+    private fun updateRouteGraphics(trackingStatus: TrackingStatus) {
+        trackingStatus.routeProgress.apply {
+            // set geometries for the route ahead and the remaining route
+            routeAheadGraphic.geometry = remainingGeometry
+            routeTraveledGraphic.geometry = traversedGeometry
+        }
+    }
+
+    /**
+     * Initialize and add route travel graphics to the map using [routeGraphic]
+     */
+    private fun createRouteGraphics(routeGraphic: Polyline) {
+        // clear any graphics from the current graphics overlay
+        mapView.graphicsOverlays[0].graphics.clear()
+
+        // create a graphic (with a dashed line symbol) to represent the route
+        routeAheadGraphic = Graphic(
+            routeGraphic,
+            SimpleLineSymbol(
+                SimpleLineSymbolStyle.Dash,
+                Color(getColor(R.color.colorPrimary)),
+                5f
+            )
+        )
+
+        // create a graphic (solid) to represent the route that's been traveled (initially empty)
+        routeTraveledGraphic = Graphic(
+            routeGraphic,
+            SimpleLineSymbol(
+                SimpleLineSymbolStyle.Solid,
+                Color.red,
+                5f
+            )
+        )
+
+        // add the graphics to the mapView's graphics overlays
+        mapView.graphicsOverlays[0].graphics.addAll(
+            listOf(
+                routeAheadGraphic,
+                routeTraveledGraphic
+            )
+        )
+    }
+
+    /**
+     * Uses Android's [voiceGuidanceText] to speak to say the latest
+     * voice guidance from the RouteTracker out loud.
+     */
+    private fun speakVoiceGuidance(voiceGuidanceText: String) {
+        if (isTextToSpeechInitialized && textToSpeech?.isSpeaking == false) {
+            textToSpeech?.speak(voiceGuidanceText, TextToSpeech.QUEUE_FLUSH, null, null)
+        }
+    }
+
+    private fun showError(message: String) {
+        Log.e(TAG, message)
+        Snackbar.make(mapView, message, Snackbar.LENGTH_SHORT).show()
+    }
+}
+
```

Comment:
Should this sample also take advantage of the `RouteTracker.isSpeechEngineReady` callback somewhere? Maybe it's not strictly needed but I think it makes sense to show off that functionality.

Reply:
Would this implementation be correct? Check if the speech is ready before collecting new `newVoiceGuidance`
```kotlin
// if route tracker speech engine is ready
if (routeTracker.isSpeechEngineReady.invoke()) {
    // listen for new voice guidance events
    lifecycleScope.launch {
        routeTracker.newVoiceGuidance.collect { voiceGuidance ->
            // use Android's text to speech to speak the voice guidance
            speakVoiceGuidance(voiceGuidance.text)
            nextDirectionTextView.text = getString(
                R.string.next_direction,
                voiceGuidance.text
            )
        }
    }
}
```

Reply:
Not quite, that property is supposed to be set with a lambda that returns true if the speech engine is ready and false if it isn't. So my first instinct would be something like
```kotlin
val routeTracker = // .. create route tracker as usual
// set a lambda which returns true if the speech engine is ready for new guidance
routeTracker.isSpeechEngineReady = { isTextToSpeechInitialized && textToSpeech?.isSpeaking == false }
```

But I see how you are interpreting it and that makes sense too (I was involved in the design PR for it so I'm already familiar with how it's supposed to be used) so I wonder if we should change the design of the callback to make it more intuitive.